[dbnode] fsync before closing checksummed files

linasm opened the pull request:
**What this PR does / why we need it**:
Adds `fsync` (`File.Sync()` call) before closing a checksummed file. This is needed to maintain transactional semantics by making fileset files durable before the checkpoint file gets written.

**Special notes for your reviewer**:
My rationale of adding this code to `fdWithDigestWriter.Close()` is that that normally `fdWithDigestWriter`s are participating in a fileset transaction and should be made durable at the moment of closing them.
An alternative approach could be to make those `Sync()` calls from `writer` code separately (before invoking `fdWithDigestWriter.Close()`), but then it would be easy to forget a `Sync()` on some of the files.

**Does this PR introduce a user-facing and/or backwards incompatible change?**:
NONE

**Does this PR require updating code package or user-facing documentation?**:
NONE

linasm committed:
[dbnode] Improve error messages on corrupt files

linasm committed:
[dbnode] fsync before closing checksummed files

linasm committed:
Merge branch 'master' into linasm/fsync-before-close-checksummed-files

arnikola commented on src/dbnode/digest/writer.go:
```diff
@@ -64,13 +64,18 @@ func (w *fdWithDigestWriter) Write(b []byte) (int, error) {
 		return err
 	}
-	return w.FdWithDigest.Close()
+	syncErr := w.FdWithDigest.Fd().Sync()
```
Do we have any data to see how this affects performance?

linasm replied:
This is a very good question. I took some real-world timings:
```
closing /var/lib/m3db/snapshots/default/250/fileset-1603432800000000000-6-info.db
flush took 13.445µs
sync took 2.545475ms
close took 83ns

closing /var/lib/m3db/snapshots/default/250/fileset-1603432800000000000-6-index.db
flush took 56.619µs
sync took 13.255645ms
close took 87ns

closing /var/lib/m3db/snapshots/default/250/fileset-1603432800000000000-6-summaries.db
flush took 73.178µs
sync took 2.316232ms
close took 82ns

closing /var/lib/m3db/snapshots/default/250/fileset-1603432800000000000-6-bloomfilter.db
flush took 34.051µs
sync took 1.641726ms
close took 79ns

closing /var/lib/m3db/snapshots/default/250/fileset-1603432800000000000-6-data.db
flush took 42.687µs
sync took 9.755168ms
close took 88ns

closing /var/lib/m3db/snapshots/default/250/fileset-1603432800000000000-6-digest.db
flush took 22.099µs
sync took 2.024822ms
close took 80ns
```
If these numbers are concerning, we could parallelize persisting of these 6 files that can be handled in any order:
https://github.com/m3db/m3/blob/20850befd7daa47faa2f0bcc10ebdbd042ec0a02/src/dbnode/persist/fs/write.go#L393-L400

linasm committed:
Rephrase error msg

linasm committed:
Merge branch 'linasm/improve-error-msgs-on-corrupt-files' into linasm/fsync-before-close-checksummed-files

linasm committed:
Also fsync checkpoint file

linasm committed:
Merge branch 'master' into linasm/fsync-before-close-checksummed-files

linasm committed:
fsync parallel and opt-in

linasm committed:
newline

linasm committed:
Merge branch 'master' into linasm/fsync-before-close-checksummed-files

linasm commented:
After our discussion:
1. Made `Sync()` calls run in parallel to offset some of the IO latency (except for the checkpoint file).
2. Made `Sync()` calls by the `fs.writer` opt-in (controlled by the new `DataWriterOpenOptions.SyncBeforeClose` option).
3. Enabled `SyncBeforeClose` for the `fs.streamingWriter`.

@robskillington which other (if any) writer use cases should have the `Sync()` enabled starting from this PR?

arnikola commented on src/dbnode/digest/writer.go:
```diff
@@ -109,3 +112,37 @@ func (w *fdWithDigestContentsWriter) WriteDigests(digests ...uint32) error {
 }
+
+func CloseAll(writers ...FdWithDigestWriter) error {
+	multiErr := xerrors.NewMultiError()
```
nit: typically we just build these as `var multiErr xerrors.MultiError`

linasm replied:
👍

arnikola commented on src/dbnode/digest/writer.go:
```diff
@@ -109,3 +112,37 @@ func (w *fdWithDigestContentsWriter) WriteDigests(digests ...uint32) error {
+
+func SyncAll(writers ...FdWithDigestWriter) error {
+	var (
+		multiErr = xerrors.NewMultiError()
```
nit: typically just `multiErr xerrors.MultiError`

linasm replied:
👍

arnikola commented on src/dbnode/digest/writer.go:
```diff
@@ -109,3 +112,37 @@ func (w *fdWithDigestContentsWriter) WriteDigests(digests ...uint32) error {
+		wg       sync.WaitGroup
+	)
+
+	wg.Add(len(writers))
+	for _, writer := range writers {
```
nit: may be slightly safer for sake of future compat to do single wg updates per loop
```suggestion
	for _, writer := range writers {
		wg.Add(1)
```

linasm replied:
👍

arnikola commented on src/dbnode/digest/writer.go:
```diff
@@ -109,3 +112,37 @@ func (w *fdWithDigestContentsWriter) WriteDigests(digests ...uint32) error {
+func SyncAll(writers ...FdWithDigestWriter) error {
+	var (
+		multiErr = xerrors.NewMultiError()
+		errMut   = sync.Mutex{}
```
nit: typically just `mu sync.Mutex` (no need to scope to errMut since no other mutexes here)

linasm replied:
👍

arnikola commented on src/dbnode/persist/fs/write.go:
```diff
@@ -390,14 +399,23 @@ func (w *writer) closeWOIndex() error {
+
+	if w.syncBeforeClose {
+		if err := digest.SyncAll(fds...); err != nil {
+			_ = digest.CloseAll(fds...)
```
nit; probably don't want to lose the error here? May be worth logging it at least, if not adding it to a full multierr

linasm replied:
Will add it to the `MultiError`.

arnikola commented on src/dbnode/persist/fs/write.go:
```diff
@@ -649,8 +668,14 @@ func writeCheckpointFile(
 		// NB(prateek): intentionally skipping fd.Close() error, as failure
 		// to write takes precedence over failure to close the file
-		fd.Close()
+		_ = fd.Close()
```
nit; Can we drop the comment and just log these errors?

linasm replied:
Will use `MultiError` here as well (and drop the comment).

arnikola reviewed (commented):
Added a few nits to code, but i'm not too familiar with the write path so not sure I can give better feedback into actual functionality; looks good from just the code perspective

linasm committed:
Merge branch 'master' into linasm/fsync-before-close-checksummed-files

linasm committed:
Address review feedback

linasm commented:
@arnikola I've addressed all of your suggestions, would appreciate if you could make one more pass on it, even if it's just from the code perspective.

arnikola commented on src/dbnode/digest/writer.go:
```diff
@@ -109,3 +112,42 @@ func (w *fdWithDigestContentsWriter) WriteDigests(digests ...uint32) error {
+func CloseAll(writers ...FdWithDigestWriter) error {
+	var multiErr xerrors.MultiError
+	for _, writer := range writers {
+		if err := writer.Close(); err != nil {
+			multiErr = multiErr.Add(err)
+		}
```
nit; `mutliErr.Add` already verifies that the error is not nil so this is not strictly necessary

arnikola commented on src/dbnode/digest/writer_test.go:
```diff
@@ -136,3 +118,40 @@ func TestFdWithDigestWriteDigestsSuccess(t *testing.T) {
+}
+
+func cleanup(fd *os.File) {
+	_ = fd.Close()
+	_ = os.Remove(fd.Name())
```
nit; worth asserting no errors here?

arnikola commented on src/dbnode/digest/writer_test.go:
```diff
@@ -136,3 +118,40 @@ func TestFdWithDigestWriteDigestsSuccess(t *testing.T) {
+
+func TestCloseAll(t *testing.T) {
+	writer, fd, _ := createTestFdWithDigestContentsWriter(t)
+	defer os.Remove(fd.Name()) // nolint:errcheck
```
nit; worth asserting no errors on these? Can probably wrap these in a function and use it for all of these nolints

arnikola reviewed (commented):
Code looks good, but you should have someone more familiar with writes give a final approval

linasm committed:
Merge branch 'master' into linasm/fsync-before-close-checksummed-files

linasm committed:
fix build
